presto-main/src/main/java/io/prestosql/sql/planner/PlanOptimizers.java

```diff
@@ -290,6 +292,8 @@ public PlanOptimizers(
                                         new PushLimitThroughMarkDistinct(),
                                         new PushLimitThroughOuterJoin(),
                                         new PushLimitThroughSemiJoin(),
+                                        new PushTopNThroughProject(),
+                                        new PushTopNThroughOuterJoin(),
```

Comment:
`PushTopNThroughOuterJoin` should be executed with
```
        builder.add(new IterativeOptimizer(
                ruleStats,
                statsCalculator,
                estimatedExchangesCostCalculator,
                ImmutableSet.of(
                        new CreatePartialTopN(),
                        new PushTopNThroughUnion())));
```

Reply:
Why move `PushTopNThroughProject` and `PushTopNThroughOuterJoin` there? Don't they belong together with `PushLimitThroughProject` and `PushLimitThroughOuterJoin`?

Reply:
> Why move PushTopNThroughProject and PushTopNThroughOuterJoin there?

It is because in that section partial `TopN` nodes are created via `CreatePartialTopN()` rule. There is already `PushTopNThroughUnion()` rule there that pushes partial aggregations though union.